[DOCS] Security best practices for GitHub Actions workflows development (#27394)
openvinotoolkit/openvino: docs/dev/ci/github_actions/security.md

# Security best practices for GitHub Actions Workflows

There are a few simple steps that we should follow to ensure our workflows are not vulnerable to common attacks:

> - Use the “permissions” key to make sure the GITHUB_TOKEN is configured with the least privileges for each job. Check [GitHub documentation](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token) on this also.

Reviewer: I would suggest to provide default value, e.g:
`permissions: read-all`

> - Use the “permissions” key to make sure the GITHUB_TOKEN is configured with the least privileges for each job. Check [GitHub documentation](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token) on this also.

Reviewer 2: ```suggestion
- Use the `permissions` key to make sure the `GITHUB_TOKEN` is configured with the least privileges for each job. Check [GitHub documentation](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/controlling-permissions-for-github_token) on this also.
```
I would also suggest adding the example from either our workflows or from the GH docs.

Author: Added

> - Declare environment variables at the step level when possible

Reviewer: As opposed to the global environment or when no environment is used at all?

> - Never use `pull_request_target` trigger event for workflows. If you want to use `pull_request_target`, contact a member of OpenVINO GitHub Actions task force first. Check [GitHub blog post](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/) on this as well.

Reviewer 2: ```suggestion
- Never use `pull_request_target` trigger event for workflows. If you want to use `pull_request_target`, contact a member of the OpenVINO GitHub Actions task force first. Check [GitHub blog post](https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/) on this as well.
```

> - Never use plain-text secrets hard-coded in GitHub Actions Workflow. If you need to use secrets, contact a member of OpenVINO GitHub Actions task force first.

Reviewer 2: ```suggestion
- Never use plain-text secrets hard-coded in GitHub Actions Workflow. If you need to use secrets, contact a member of the OpenVINO GitHub Actions task force first.
```

- Be careful with user input.

>     Most of GitHub context variables propagated from user input. That means they should be treated as an untrusted and potentially malitious. There are some tactics you can use to mitigate the risk:

Reviewer 2: ```suggestion
    Most GitHub context variables are propagated from user input. That means they should be treated as untrusted and potentially malicious. There are some tactics you can use to mitigate the risk:
```

Author: Oops 😅 thanks a lot!

    - Instead of using inline scripts, create an action and pass the variable as an argument
    - Put the value into an environment variable for the step, and use the variable in the script

    More details are available in [this](https://securitylab.github.com/resources/github-actions-untrusted-input/) blog post.

> - When using third-party actions, pin the version with a commit hash rather than a tag to shield your workflow from potential supply-chain compromise. 

Reviewer: Let's insert an example:
`uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2`

> - When using third-party actions, pin the version with a commit hash rather than a tag to shield your workflow from potential supply-chain compromise. 

Reviewer 2: I would suggest adding an example from out workflows.
